cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -362,3 +384,144 @@ func (m *McmManager) GetMachineDeploymentNodes(machinedeployment *MachineDeploym
 	}
 	return nodes, nil
 }
+
+//GetMachineDeploymentNodeTemplate returns the NodeTemplate which belongs to the MachineDeployment.
+func (m *McmManager) GetMachineDeploymentNodeTemplate(machinedeployment *MachineDeployment) (*nodeTemplate, error) {
+
+	md, err := m.machineclient.MachineDeployments(m.namespace).Get(machinedeployment.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("Unable to fetch MachineDeployment object %s, Error: %v", machinedeployment.Name, err)
+	}
+
+	var region string
+	var tags map[string]string
+	var instance instanceType
+	machineClass := md.Spec.Template.Spec.Class
+	switch machineClass.Kind {
+	case "AWSMachineClass":
+		mc, err := m.machineclient.AWSMachineClasses(m.namespace).Get(machineClass.Name, metav1.GetOptions{})
```

Comment:
Ideally, we should avoid bringing cloud-provider specific knowledge into the neutral mcm-provider. But unfortunately, we still have cloud-provider specific MachineClasses in MCM, hence this would be needed here. 
  -  With the out of tree efforts, we shall have only one type `MachineClass`, which will then have subfield for cloud-provider-ref. [#178](https://github.com/gardener/machine-controller-manager/issues/178)

For now, let's keep it the provider-specific way. Can you please include all the providers supported by MCM? Eg. AWS, GCP, Azure, Alicloud, Packet, and Openstack

Reply:
We need to rebase from upstream if we want to add all providers. 
I think we should do it.

Reply:
Yes, we should either rebase autoscaler or find a way to include support for other providers. For now, we might want to go ahead reviewing other parts and merge this PR with the support of AWS and Azure only.  
- As I can see, the only provider-specific dependency is `instance types`. Can we think of a solution, where we could support other providers via MCM - ones not supported by even upstream autoscaler? Something, as vague as bringing `instanceType` files of different providers into `mcm-provider`, and enable scale-to-zero from there. 
   - That could be useful in case auto scaler decides to go out-of-tree for cloud-providers in future. 
Wdyt ?

Reply:
@hardikdr @rewiko wouldn't it be reasonable to rather let each provider class controller to run this provider specific logic to fetch values from a type, and let the class expose them via status/annotations, so we can keep provider neutrality here by using a dynamic client to fetch the class and retrieve the values from the known provider agnostic status/annotations

Reply:
Thanks @enxebre, this is an excellent suggestion, we have also been thinking in similar lines earlier. We shall initiate designing/implementing the solution in this direction soon at MCM, will keep you in the loop. 
Btw what do you think about the comment [below](https://github.com/gardener/autoscaler/pull/26#issuecomment-581567741).

Reply:
Thanks @hardikdr @rewiko, sounds great. Another thing to take into consideration would be how to handle existing clusters. I'm not sure how gardener handle upgrades but fwiw as soon as this is shipped into an existing cluster it will enable scaling from/to zero for existing nodeGroups which might be under the hard assumption that minimum=1. So might want to consider to make this feature opt-in some how or providing an explicit transitional path for the user.

Reply:
@enxebre it was not working anyway when min set to zero because it couldn't build the nodeTemplate, so I think that will be a non disruptive change.

Reply:
Yes, this should not disrupt the existing clusters, it will only enable the feature for new worker-pools with min=0.